core-processor/src/ext.rs

```diff
@@ -915,6 +916,16 @@ impl EnvExt for Ext {
         Ok((mid, pid))
     }
 
+    fn reply_deposit(&mut self, message_id: MessageId, amount: u64) -> Result<(), Self::Error> {
+        self.charge_gas_if_enough(amount)?;
```

Comment:
shouldn't we `reduce` instead of `burn`? The `amount` will be re-used for the other message.